code/controllers/subsystems/news.dm

```diff
@@ -0,0 +1,154 @@
+
+/datum/controller/subsystem/news
+	name = "News"
+	flags = SS_NO_FIRE
+	var/list/datum/feed_channel/network_channels = list()
+	var/datum/feed_message/wanted_issue
+
+/datum/controller/subsystem/news/Recover()
+	src.network_channels = SSnews.network_channels
+	src.wanted_issue = SSnews.wanted_issue
+
+/datum/controller/subsystem/news/New()
+	NEW_SS_GLOBAL(SSnews)
+
+/datum/controller/subsystem/news/Initialize(timeofday)
+	CreateFeedChannel("Station Announcements", "SS13", 1, 1, "New Station Announcement Available")
+	if(config.sql_enabled)
+		load_from_sql()
+	..()
+
+/datum/controller/subsystem/news/proc/load_from_sql()
+	if(!establish_db_connection(dbcon))
+		log_debug("SSnews: SQL ERROR - Failed to connect.")
+		return
+	var/DBQuery/channel_query = dbcon.NewQuery("SELECT id, name, author, locked, is_admin_channel, announcement FROM ss13_news_channels WHERE deleted_at IS NULL ORDER BY name ASC")
+	channel_query.Execute()
+	while(channel_query.NextRow())
+		CHECK_TICK
+		try
+			CreateFeedChannel(
+				channel_query.item[2],
+				channel_query.item[3],
+				text2num(channel_query.item[4]),
+				text2num(channel_query.item[5]),
+				channel_query.item[6])
+		catch(var/exception/ec)
+			log_debug("SSnews: Error when loading channel: [ec]")
+		CHECK_TICK
+		var/DBQuery/news_query = dbcon.NewQuery("SELECT body, author, is_admin_message, message_type, time_stamp FROM ss13_news_stories WHERE deleted_at IS NULL AND channel_id = :channel_id: ORDER BY created_at DESC")
+		news_query.Execute(list("channel_id" = channel_query.item[1]))
+		while(news_query.NextRow())
+			try
+				SubmitArticle(news_query.item[1], news_query.item[2], channel_query.item[2], null, text2num(news_query.item[3]), news_query.item[4], news_query.item[5])
+			catch(var/exception/en)
+				log_debug("SSnews: Error when loading news: [en]")
+
+
+
+/datum/controller/subsystem/news/proc/CreateFeedChannel(var/channel_name, var/author, var/locked, var/adminChannel = 0, var/announcement_message)
+	var/datum/feed_channel/newChannel = new /datum/feed_channel
+	newChannel.channel_name = channel_name
+	newChannel.author = author
+	newChannel.locked = locked
+	newChannel.is_admin_channel = adminChannel
+	if(announcement_message)
+		newChannel.announcement = announcement_message
+	else
+		newChannel.announcement = "Breaking news from [channel_name]!"
+	network_channels += newChannel
+
+/datum/controller/subsystem/news/proc/SubmitArticle(var/msg, var/author, var/channel_name, var/obj/item/weapon/photo/photo, var/adminMessage = 0, var/message_type = "", var/time_stamp)
+	var/datum/feed_message/newMsg = new /datum/feed_message
+	newMsg.author = author
+	newMsg.body = msg
+	if(time_stamp)
+		newMsg.time_stamp = time_stamp
+	else
+		newMsg.time_stamp = "[worldtime2text()]"
+	newMsg.is_admin_message = adminMessage
+	if(message_type)
+		newMsg.message_type = message_type
+	if(photo)
+		newMsg.img = photo.img
+		newMsg.caption = photo.scribble
+	for(var/datum/feed_channel/FC in network_channels)
```

Comment:
I wonder if it'd be better to avoid this whenever possible. ie send the channel as an optional argument, and only do this search if there's a channel name but no object? The idea being that while loading stuff, this loop is executed quite a few times. Despite there being no need for it.

Or just make _this_ require a channel object, provide a small proc for `SSnews.GetChannel(channel_name)`.

_Or_ make network_channels be a key-value list. Instead of just a plain list. This'll make the search speed not really worth considering.